### ISSUE TITLE
SplineTransformer

### ISSUE THREAD

**Issue by MaxHalford** [online-ml/river#1367]
[See scikit-learn](https://scikit-learn.org/stable/modules/generated/sklearn.preprocessing.SplineTransformer.html?highlight=spline#sklearn.preprocessing.SplineTransformer)

**Comment by sarim-zafar** [comment]
Generated this script using an LLM ... thought it would be a nice starting point for future reference.

```
import numpy as np
from river import base
from scipy.interpolate import BSpline

class SplineTransformer(base.Transformer):
    def __init__(self, n_knots=5, degree=3, knots='uniform', extrapolation='constant', include_bias=True, order='C'):
        self.n_knots = n_knots
        self.degree = degree
        self.knots = knots
        self.extrapolation = extrapolation
        self.include_bias = include_bias
        self.order = order
        self.bsplines_ = []
        self.fitted = False

    def fit_one(self, x):
        X = np.array(list(x.values())).reshape(1, -1)
        n_samples, n_features = X.shape

        if isinstance(self.knots, str):
            base_knots = self._get_base_knot_positions(X, n_knots=self.n_knots, knots=self.knots)
        else:
            base_knots = np.array(self.knots, dtype=np.float64)
        
        n_knots = base_knots.shape[0]
        n_splines = n_knots + self.degree - 1 if self.extrapolation != 'periodic' else n_knots - 1
        degree = self.degree

        if self.extrapolation == 'periodic':
            period = base_knots[-1] - base_knots[0]
            knots = np.r_[
                base_knots[-(degree + 1):-1] - period,
                base_knots,
                base_knots[1:(degree + 1)] + period,
            ]
        else:
            dist_min = base_knots[1] - base_knots[0]
            dist_max = base_knots[-1] - base_knots[-2]
            knots = np.r_[
                np.linspace(base_knots[0] - degree * dist_min, base_knots[0] - dist_min, num=degree),
                base_knots,
                np.linspace(base_knots[-1] + dist_max, base_knots[-1] + degree * dist_max, num=degree),
            ]

        coef = np.eye(n_splines, dtype=np.float64)
        if self.extrapolation == 'periodic':
            coef = np.concatenate((coef, coef[:degree, :]))

        extrapolate = self.extrapolation in ["periodic", "continue"]
        self.bsplines_ = [
            BSpline.construct_fast(knots[:, i], coef, self.degree, extrapolate=extrapolate)
            for i in range(n_features)
        ]
        
        self.fitted = True
        return self

    def transform_one(self, x):
        if not self.fitted:
            raise RuntimeError("The SplineTransformer instance is not fitted yet.")
        
        X = np.array(list(x.values())).reshape(1, -1)
        n_samples, n_features = X.shape
        n_splines = self.bsplines_[0].c.shape[1]

        XBS = np.zeros((n_samples, n_features * n_splines), dtype=np.float64, order=self.order)

        for i in range(n_features):
            spl = self.bsplines_[i]
            xmin, xmax = spl.t[self.degree], spl.t[-self.degree - 1]
            f_min, f_max = spl(xmin), spl(xmax)
            mask = (xmin <= X[:, i]) & (X[:, i] <= xmax)

            XBS[:, (i * n_splines):((i + 1) * n_splines)] = spl(X[:, i])
        
        if self.include_bias:
            return dict(enumerate(XBS.flatten(), 1))
        else:
            XBS = XBS[:, :-1]
            return dict(enumerate(XBS.flatten(), 1))

    def _get_base_knot_positions(self, X, n_knots=10, knots="uniform"):
        if knots == "quantile":
            percentiles = 100 * np.linspace(start=0, stop=1, num=n_knots, dtype=np.float64)
            knots = np.percentile(X, percentiles, axis=0)
        else:
            x_min = np.amin(X, axis=0)
            x_max = np.amax(X, axis=0)
            knots = np.linspace(start=x_min, stop=x_max, num=n_knots, endpoint=True, dtype=np.float64)
        return knots

```import numpy as np
from river import base
from scipy.interpolate import BSpline

class SplineTransformer(base.Transformer):
    def __init__(self, n_knots=5, degree=3, knots='uniform', extrapolation='constant', include_bias=True, order='C'):
        self.n_knots = n_knots
        self.degree = degree
        self.knots = knots
        self.extrapolation = extrapolation
        self.include_bias = include_bias
        self.order = order
        self.bsplines_ = []
        self.fitted = False

    def fit_one(self, x):
        X = np.array(list(x.values())).reshape(1, -1)
        n_samples, n_features = X.shape

        if isinstance(self.knots, str):
            base_knots = self._get_base_knot_positions(X, n_knots=self.n_knots, knots=self.knots)
        else:
            base_knots = np.array(self.knots, dtype=np.float64)
        
        n_knots = base_knots.shape[0]
        n_splines = n_knots + self.degree - 1 if self.extrapolation != 'periodic' else n_knots - 1
        degree = self.degree

        if self.extrapolation == 'periodic':
            period = base_knots[-1] - base_knots[0]
            knots = np.r_[
                base_knots[-(degree + 1):-1] - period,
                base_knots,
                base_knots[1:(degree + 1)] + period,
            ]
        else:
            dist_min = base_knots[1] - base_knots[0]
            dist_max = base_knots[-1] - base_knots[-2]
            knots = np.r_[
                np.linspace(base_knots[0] - degree * dist_min, base_knots[0] - dist_min, num=degree),
                base_knots,
                np.linspace(base_knots[-1] + dist_max, base_knots[-1] + degree * dist_max, num=degree),
            ]

        coef = np.eye(n_splines, dtype=np.float64)
        if self.extrapolation == 'periodic':
            coef = np.concatenate((coef, coef[:degree, :]))

        extrapolate = self.extrapolation in ["periodic", "continue"]
        self.bsplines_ = [
            BSpline.construct_fast(knots[:, i], coef, self.degree, extrapolate=extrapolate)
            for i in range(n_features)
        ]
        
        self.fitted = True
        return self

    def transform_one(self, x):
        if not self.fitted:
            raise RuntimeError("The SplineTransformer instance is not fitted yet.")
        
        X = np.array(list(x.values())).reshape(1, -1)
        n_samples, n_features = X.shape
        n_splines = self.bsplines_[0].c.shape[1]

        XBS = np.zeros((n_samples, n_features * n_splines), dtype=np.float64, order=self.order)

        for i in range(n_features):
            spl = self.bsplines_[i]
            xmin, xmax = spl.t[self.degree], spl.t[-self.degree - 1]
            f_min, f_max = spl(xmin), spl(xmax)
            mask = (xmin <= X[:, i]) & (X[:, i] <= xmax)

            XBS[:, (i * n_splines):((i + 1) * n_splines)] = spl(X[:, i])
        
        if self.include_bias:
            return dict(enumerate(XBS.flatten(), 1))
        else:
            XBS = XBS[:, :-1]
            return dict(enumerate(XBS.flatten(), 1))

    def _get_base_knot_positions(self, X, n_knots=10, knots="uniform"):
        if knots == "quantile":
            percentiles = 100 * np.linspace(start=0, stop=1, num=n_knots, dtype=np.float64)
            knots = np.percentile(X, percentiles, axis=0)
        else:
            x_min = np.amin(X, axis=0)
            x_max = np.amax(X, axis=0)
            knots = np.linspace(start=x_min, stop=x_max, num=n_knots, endpoint=True, dtype=np.float64)
        return knots

**Comment by kulbachcedric** [comment]
👋 Hi @sarim-zafar,

we’re currently doing a cleanup of older Issues in the *river* repository and wanted to check in.

Is this Issue still relevant and something you’re actively working on or would like to see addressed?

If we don’t hear back by **January 31**, we’ll assume it’s no longer active and close it to keep the tracker tidy — but of course you’re very welcome to reply later or reopen it if needed 🙂

Feel free to create a PR with your already created code ;-) 

Thanks a lot!
